### PR TITLE
feat: update polling ux and incorporate api changes

### PR DESCRIPTION
**Pull request by nikhilvenkatraman-okta**
Resolves: OKTA-289354

## Description:

- Updates pollController to match changes in API contract 
- Updates polling UX to add spinner between each poll

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

### Screenshot/Video:

https://okta.box.com/s/qlmxbu0zsb8999i4ka8bkfbpwc0e9956

### Issue:

- [OKTA-289354](https://oktainc.atlassian.net/browse/OKTA-289354)

@magizh-okta @aarongranick-okta @taylorlaubach-okta @vincentvoong-okta

### CODE REVIEW

File: src/PollController.js [hunk 1]
```diff
@@ -78,8 +94,8 @@ function (Okta, FormController) {
       },
 
       initialize: function (options) {
-        this.transaction = options.appState.get('transaction');
-        this.factorPollingInterval = this.transaction.factor.profile.refresh;
+        this.transactionObject = options.appState.get('transaction');
+        this.factorPollingInterval = this.transactionObject.transaction.profile.refresh;
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:98]: yeah, the API is sending field as `transaction`. I had asked @petermiller-okta about this.

File: src/PollController.js [hunk 2]
```diff
@@ -42,23 +43,38 @@ function (Okta, FormController) {
...
+              if (factorPollingIntervalSeconds === 0) {
+                /* when countdown hits 0
+                 - stop current poll
+                 - show spinner
+                 - start next poll
+                */
                 this._stopPolling();
-                this._startPolling();
+                this.$('.okta-waiting-spinner').show();
+                setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:75]: Yes, this is just purely visual. In the previous UI, every time we re-started the poll the counter would go down and just restart back up again. We received some feedback that this was confusing to endusers  and so we are showing spinner briefly between subsequent poll requests.

[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:75]: The spinner give the user feedback that the current poll has finished and that we are re-starting a new polling request.

File: src/PollController.js [hunk 3]
```diff
@@ -42,23 +43,38 @@ function (Okta, FormController) {
...
+        })
+      ],
       _startPolling: function () {
         // start polling request 
-        this.transaction.poll()
+        this.transactionObject.poll()
           .then((resp) => {
-            this.options.appState.set('transaction', resp);
-            const factorPollingInterval = resp.factor.profile.refresh;
+            this.options.appState.trigger('setTransaction', resp);
```
[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:58]: the `transaction` is just a field that is being sent in the response.
I am actually going to make a small update here. One thing @aarongranick-okta pointed out to me was that we probably don't need to set this new transaction if our status has not changed, to prevent a re-render.

We need to set the transaction so that when this changes to success we can proceed and render a different controller accordingly.

TLDR: I will update this to 

```  
 if (resp.status === 'SUCCESS') {
              this.$('.okta-waiting-spinner').hide();
              this.options.appState.set('transaction', resp);
            } 
```

File: src/PollController.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (magizh-okta) [reply src/PollController.js:78]: yeah but why 200 ms delay ?

[the remark this answers is not in the document]
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:78]: Its an arbitrary value, just enough time to see the spinner.

File: src/PollController.js [hunk 4]
```diff
@@ -51,41 +51,52 @@ function (Okta, FormController, FormType) {
...
-                this.$('.okta-waiting-spinner').show();
-                setTimeout(() => {
-                  this._startPolling();
-                }, 200);
-              } else {
-                // reduce countdown after every second
-                factorPollingIntervalSeconds = factorPollingIntervalSeconds-1;
-              }
-            }, 1000);
+            if (resp.data && resp.status !== 'POLL') {
```
Comment (nikhilvenkatraman-okta) [inline src/PollController.js:58]: Waiting to get confirmation from backend on this check to see if we can only get into SUCCESS state after poll, if so I will update this.

File: src/PollController.js [hunk 5]
```diff
@@ -42,34 +43,60 @@ function (Okta, FormController) {
...
+        this.countDown  = setInterval(() => {
+          // update title after every second and check if countdown == 1 to poll again
+          this._updateTitle(factorPollingIntervalSeconds);
+          if (factorPollingIntervalSeconds === 0) {
+            /* when countdown hits 0
+                         - stop current poll
+                         - show spinner
+                         - start next poll
+                        */
+            this._stopPolling();
```
Comment (magizh-okta) [inline src/PollController.js:82]: I would name this as _stopCountDown to avoid confusion.

File: src/PollController.js [hunk 6]
```diff
@@ -42,34 +43,60 @@ function (Okta, FormController) {
...
+            this.$('.okta-waiting-spinner').show();
+            // start after a small delay so that the spinner does not get hidden too soon
+            setTimeout(() => {
+              this._doPoll();
```
Comment (magizh-okta) [inline src/PollController.js:86]: Lets name it check status as this function doesnt actually poll.

File: src/PollController.js [hunk 7]
```diff
@@ -42,34 +43,60 @@ function (Okta, FormController) {
...
       },
 
+      _startCountDown: function (factorPollingIntervalSeconds) {
+        // start one second countdown for next poll request
+        this.countDown  = setInterval(() => {
+          // update title after every second and check if countdown == 1 to poll again
+          this._updateTitle(factorPollingIntervalSeconds);
+          if (factorPollingIntervalSeconds === 0) {
+            /* when countdown hits 0
+                         - stop current poll
```
Comment (magizh-okta) [inline src/PollController.js:78]: nit: Modify comments to say 

   - stop current countDown.
   - show spinner 
    - Check status

File: src/PollController.js [hunk 8]
```diff
@@ -78,14 +105,15 @@ function (Okta, FormController) {
...
-        this.title = Okta.loc('polling.title','login', [factorPollingIntervalSeconds]);
+        this.transactionObject = options.appState.get('transaction');
+        this.factorPollingIntervalSeconds = Math.ceil(this.transactionObject.transaction.profile.refresh/1000);
+        this._updateTitle(this.factorPollingIntervalSeconds);
       },
   
       postRender: function () {
-        this._startPolling(this.factorPollingInterval);
+        setTimeout(() => {
+          this._startCountDown(this.factorPollingIntervalSeconds);
```
Comment (magizh-okta) [inline src/PollController.js:115]: I guess we dont need the setTimeout here as the startCountDown takes care of calling doPoll after the factorPollingIntervalSeconds anyway.

File: test/unit/spec/PollController_spec.js [hunk 9]
```diff
@@ -72,47 +72,49 @@ function (Okta, Q, OktaAuth, Util, PollingForm, Expect, Router, $sandbox, resPol
...
-            // 3rd poll
-            Expect.isJsonPost($.ajax.calls.argsFor(3), {
-              url: 'https://example.okta.com/api/v1/authn/factors/okta-poll/poll',
-              data: {
-                stateToken: '00_J1qxqyLs-6ZutUUWfbqm-1nqnW6n2o5z2wnBRHs'
-              }
-            });
-          });
+            done();
+          }, 9000);
```
Comment (magizh-okta) [inline test/unit/spec/PollController_spec.js:111]: I am guessing we can reach 3 calls in 3 seconds itself. 

1) First call - takes us to poll screen 
2) Second call happens after timeout specificed ( 1sec) 
3) Third call happens after 1200ms. 
So in 2200ms 3 calls should have been made right.

File: test/unit/spec/PollController_spec.js [hunk 10]
```diff
@@ -44,21 +42,23 @@ function (Okta, Q, OktaAuth, Util, PollingForm, Expect, Router, $sandbox, resPol
...
-    describe('PollingForm Content', function () {
-      itp('shows the correct content on load', function () {
+    describe('Form Content', function () {
+      it('shows the correct content on load', function (done) {
         return setup().then(function (test) {
-          const title = 'There are too many users trying to sign in right now. We will automatically retry in 2 seconds.';
+          const title = 'There are too many users trying to sign in right now. We will automatically retry in 1 seconds.';
           expect(test.form.pageTitle().text().trim()).toBe(title);
-        });
+          done();
```
Comment (magizh-okta) [inline test/unit/spec/PollController_spec.js:50]: Why are we using done here ? There is nothing async here right.

File: test/unit/spec/PollController_spec.js [hunk 11]
```diff
@@ -44,21 +42,23 @@ function (Okta, Q, OktaAuth, Util, PollingForm, Expect, Router, $sandbox, resPol
...
+        }).catch(done.fail);
       });
-      itp('has the cancel button', function () {
+      it('has the cancel button', function (done) {
         return setup().then(function (test) {
           expect(test.form.cancelButton()).toExist();
           expect(test.form.cancelButton().attr('value')).toBe('Cancel');
           expect(test.form.cancelButton().attr('class')).toBe('button button-primary');
-        });
+          done();
```
Comment (magizh-okta) [inline test/unit/spec/PollController_spec.js:58]: same here.

File: test/unit/spec/PollController_spec.js [hunk 12]
```diff
@@ -72,47 +72,49 @@ function (Okta, Q, OktaAuth, Util, PollingForm, Expect, Router, $sandbox, resPol
...
-          return Expect.wait(function () {
-            return $.ajax.calls.count() > 3;
-          }, test);
-        })
-          .then(function () {
+  Expect.describe('Polling API', function () {
+    describe('called on load', function () {
+      var originalTimeout;
+      beforeEach(function () {
+        originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
```
Comment (magizh-okta) [inline test/unit/spec/PollController_spec.js:87]: What does this do. Do we need this when we are not using jasmie.clock /.

File: test/unit/spec/PollController_spec.js (hunk 10 above)
Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js:50]: these tests were continuing to poll and make network requests after it ended. I wanted some way for these to stop polling after the test is over.

File: test/unit/spec/PollController_spec.js (hunk 9 above)
Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js:111]: In theory yes, but I had to play around with it. I can update this to 5000 as thats the earliest it ended.

File: src/PollController.js (hunk 4 above)
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:58]: going to leave this as is

File: test/unit/spec/PollController_spec.js (hunk 10 above)
Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js:50]: https://jasmine.github.io/tutorials/a

File: test/unit/spec/PollController_spec.js (hunk 12 above)
Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js:87]: the default timeout for each test is 10000, this is just giving it some more time to execute.

Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js]: I managed to remove these 🤞

File: test/unit/spec/PollController_spec.js [hunk 13]
```diff
@@ -108,13 +102,8 @@ function (Okta, OktaAuth, Util, PollingForm, Expect, Router, $sandbox, resPollin
               }
             });
             done();
-          }, 9000);
+          }, 5000);
```
Comment (haishengwu-okta) [inline test/unit/spec/PollController_spec.js:105]: certainly not idea to use real timeout in test. what's the challenge to mock time?

Comment (nikhilvenkatraman-okta) [reply test/unit/spec/PollController_spec.js:105]: @haishengwu-okta I have tried using jasmine.clock(). If we do that we'll see this error. 
```  
debug.js:21 Error: Jasmine Clock was unable to install over custom global timer functions. Is the clock already installed?
    at <Jasmine>
    at UserContext.<anonymous> (PollController_spec.js:91)
    at <Jasmine>

```
uninstalling before installing the clock also does not resolve the problem.

@aarongranick-okta has a current PR to fix these issues https://github.com/okta/okta-signin-widget/compare/4.0?expand=1#diff-8ef46a6d97cdf70fb7322bc2ef22b3f8R340 and I was hoping to avoid making all these changes here

Comment (haishengwu-okta) [reply test/unit/spec/PollController_spec.js:105]: oh, I see. Thanks. Looks like it might be some mess of Jasmine Clock somewhere but yeah, it's hard to figure it out.

File: src/PollController.js [hunk 14]
```diff
@@ -40,52 +41,77 @@ function (Okta, FormController) {
...
+          }
+        }, 1000);
+      },
+
+      _updateTitle: function (factorPollingIntervalSeconds) {
+        this.title = Okta.loc('polling.title','login', [factorPollingIntervalSeconds]);
+        this.$el.find('.okta-form-title').text(this.title);
+      },
+
+      _stopCountDown: function () {
```
Comment (aarongranick-okta) [inline src/PollController.js:100]: _stopCountdown should also clear the timeout set on line 85

Comment (nikhilvenkatraman-okta) [reply src/PollController.js:100]: fixed

**Comment by aarongranick-okta** [comment]
related issue: OKTA-291740

File: src/PollController.js [hunk 15]
```diff
@@ -40,52 +41,81 @@ function (Okta, FormController) {
...
-        const factorPollingIntervalSeconds = Math.floor(this.factorPollingInterval/1000);
-        this.title = Okta.loc('polling.title','login', [factorPollingIntervalSeconds]);
+        this.transactionObject = options.appState.get('transaction');
+        this.factorPollingIntervalSeconds = Math.ceil(this.transactionObject.transaction.profile.refresh/1000);
+        this._updateTitle(this.factorPollingIntervalSeconds);
       },
   
       postRender: function () {
-        this._startPolling(this.factorPollingInterval);
+        this._startCountDown(this.factorPollingIntervalSeconds);
```
Comment (magizh-okta) [inline src/PollController.js:118]: Any reason not to do it in initialize ?

**Comment by nikhilvenkatraman-okta** [comment]
Like Aaron pointed out https://oktainc.atlassian.net/browse/OKTA-291740 is the JIRA for refactor

File: src/PollController.js (hunk 15 above)
Comment (nikhilvenkatraman-okta) [reply src/PollController.js:118]: fixed
